mysql/mysqldump_test.go

```diff
@@ -0,0 +1,561 @@
+}
+
+func TestProcessMySQLDump_Rows(t *testing.T) {
+	conv, _ := runProcessMySQLDump("CREATE TABLE cart (a text, n bigint);\n" +
+		"INSERT INTO cart (a, n) VALUES ('a42', 2);")
+	assert.Equal(t, int64(1), conv.Rows())
+}
+
+func TestProcessMySQLDump_BadRows(t *testing.T) {
+	conv, _ := runProcessMySQLDump("CREATE TABLE cart (a text, n bigint);\n" +
+		"INSERT INTO cart (a, n) VALUES ('a42', 'not_a_number');")
+	assert.Equal(t, int64(1), conv.BadRows())
+}
+
+func TestProcessMySQLDump_GetBadRows(t *testing.T) {
+	conv, _ := runProcessMySQLDump("CREATE TABLE cart (a text, n bigint);\n" +
+		"INSERT INTO cart (a, n) VALUES ('a42', 'not_a_number');")
+	assert.Equal(t, 1, len(conv.SampleBadRows(100)))
+}
+
+func TestProcessMySQLDump_AddPrimaryKeys(t *testing.T) {
+	cases := []struct {
+		name           string
+		input          string
+		expectedSchema map[string]ddl.CreateTable
+	}{
+		{
+			name:  "Shopping cart",
+			input: "CREATE TABLE cart (productid text, userid text, quantity bigint);",
+			expectedSchema: map[string]ddl.CreateTable{
+				"cart": ddl.CreateTable{
+					Name:     "cart",
+					ColNames: []string{"productid", "userid", "quantity", "synth_id"},
+					ColDefs: map[string]ddl.ColumnDef{
+						"productid": ddl.ColumnDef{Name: "productid", T: ddl.String{Len: ddl.MaxLength{}}},
+						"userid":    ddl.ColumnDef{Name: "userid", T: ddl.String{Len: ddl.MaxLength{}}},
+						"quantity":  ddl.ColumnDef{Name: "quantity", T: ddl.Int64{}},
+						"synth_id":  ddl.ColumnDef{Name: "synth_id", T: ddl.Int64{}},
+					},
+					Pks: []ddl.IndexKey{ddl.IndexKey{Col: "synth_id"}}}},
+		},
+		{
+			name:  "synth_id clash",
+			input: "CREATE TABLE test (synth_id text, synth_id0 text, synth_id1 bigint);",
+			expectedSchema: map[string]ddl.CreateTable{
+				"test": ddl.CreateTable{
+					Name:     "test",
+					ColNames: []string{"synth_id", "synth_id0", "synth_id1", "synth_id2"},
+					ColDefs: map[string]ddl.ColumnDef{
+						"synth_id":  ddl.ColumnDef{Name: "synth_id", T: ddl.String{Len: ddl.MaxLength{}}},
+						"synth_id0": ddl.ColumnDef{Name: "synth_id0", T: ddl.String{Len: ddl.MaxLength{}}},
+						"synth_id1": ddl.ColumnDef{Name: "synth_id1", T: ddl.Int64{}},
+						"synth_id2": ddl.ColumnDef{Name: "synth_id2", T: ddl.Int64{}},
+					},
+					Pks: []ddl.IndexKey{ddl.IndexKey{Col: "synth_id2"}}}},
+		},
+	}
+	for _, tc := range cases {
+		conv, _ := runProcessMySQLDump(tc.input)
+		conv.AddPrimaryKeys()
+		if tc.expectedSchema != nil {
+			assert.Equal(t, tc.expectedSchema, stripSchemaComments(conv.SpSchema), tc.name+": Schema did not match")
+		}
+	}
+}
+
+func runProcessMySQLDump(s string) (*internal.Conv, []spannerData) {
+	conv := internal.MakeConv()
+	conv.SetLocation(time.UTC)
+	conv.SetSchemaMode()
+	ProcessMySQLDump(conv, internal.NewReader(bufio.NewReader(strings.NewReader(s)), nil))
+	conv.SetDataMode()
+	var rows []spannerData
+	conv.SetDataSink(
+		func(table string, cols []string, vals []interface{}) {
```

Comment:
This function can be on the same level as `SetDataSink`. No need to break them into two lines.